core/sql/src/main/java/org/trafodion/sql/HTableClient.java

```diff
@@ -421,29 +839,140 @@ else if (versions > 0)
+		    	  }
+		      }	    	
+		    }
+		    if (columnsToRemove != null)
+		    { //we are almost done checking if Exclude version of SingleColumnnValueFilter can be used. Th elast check s about to know if there is a IS_NULL_NULL
+		      //operation that cannot be using the Exclude method, as it is transformed in a filterList with OR, therefore we cannot guaranty that the SingleColumnValueExcludeFilter
+		      //performing the exclusion will be reached.
+		    	boolean is_null_nullFound = false;
+		    	for (Object o:compareOpList ){
+		    		if (new String((byte[])o).equals("IS_NULL_NULL")){
+		    			is_null_nullFound = true;
+		    			break;
+		    		}		    			
+		    	}
+		    	if (is_null_nullFound){
+		    		columnsToRemove = null; // disable Exclude method version of SingleColumnnValueFilter
+		    	}else
+		    		narrowDownResultColumns = false; // we will use the Exclude version of SingleColumnnValueFilter, so bypass the Family/QualifierFilter method
+		    }
+		    Filter f =constructV2Filter(colNamesToFilter,compareOpList,colValuesToCompare, columnsToRemove);
+		    if (f==null) return false; // error logging done inside constructV2Filter
+		    list.addFilter(f);
+		}//end V2
+		else{// deal with V1
+		    list = new FilterList(FilterList.Operator.MUST_PASS_ALL);
+		    
+		    for (int i = 0; i < colNamesToFilter.length; i++) {
+		      byte[] colName = (byte[])colNamesToFilter[i];
+		      byte[] coByte = (byte[])compareOpList[i];
+		      byte[] colVal = (byte[])colValuesToCompare[i];
+	
+		      if ((coByte == null) || (colVal == null)) {
+		        return false;
+		      }
+		      String coStr = new String(coByte);
+		      CompareOp co = CompareOp.valueOf(coStr);
+	
+		      SingleColumnValueFilter filter1 = 
+		          new SingleColumnValueFilter(getFamily(colName), getName(colName), 
+		              co, colVal);
+		      list.addFilter(filter1);
+		    }			
+		}//end V1
+	    // if we added a column for predicate eval, we need to filter down result columns
+	    FilterList resultColumnsOnlyFilter = null;
+	    if (narrowDownResultColumns){  	    	
+  		    HashMap<String,ArrayList<byte[]>> hm = new HashMap<String,ArrayList<byte[]>>(3);//use to deal with multiple family table
+  		    // initialize hm with list of columns requested for output
+  		    	for (int i=0; i<columns.length; i++){ // if we are here we know columns is not null
+  		    		if (hm.containsKey(new String(getFamily((byte[])columns[i])))){
+  		    			hm.get(new String(getFamily((byte[])columns[i]))).add((byte[])columns[i]);
+  		    		}else{
+  		    			ArrayList<byte[]> al = new ArrayList<byte[]>();
+  		    			al.add((byte[])columns[i]);
+  		    			hm.put(new String(getFamily((byte[])columns[i])), al);
+  		    		}	    			
+  		    	}
+  		    	
+  	    	if (hm.size()==1){//only one column family
+  	    		resultColumnsOnlyFilter = new FilterList(FilterList.Operator.MUST_PASS_ALL);
```

Comment:
Hi,

Morning.

I have worked on the separation of the original extractHbaseFilterPredsV2()
into two routines: isFeasibleToExtractHbaseFilterPredsV2()
and extractHbaseFilterPredsV2(). Please review the attached two files.

While this realizes the objective of one function does one thing only, I
still feel both new functions can be simplified further, in that there is
still striking level of similarities between the AND and OR branches.
Sounds like the two branches can be reduced into 1.

Thanks --Qifan



On Tue, Jan 12, 2016 at 4:41 PM, Hans Zeller <notifications@github.com>
wrote:

> In core/sql/src/main/java/org/trafodion/sql/HTableClient.java
> <https://github.com/apache/incubator-trafodion/pull/255#discussion_r49526031>
> :
>
> > +	    FilterList resultColumnsOnlyFilter = null;
> > +	    if (narrowDownResultColumns){  	    	
> > +  		    HashMap<String,ArrayList<byte[]>> hm = new HashMap<String,ArrayList<byte[]>>(3);//use to deal with multiple family table
> > +  		    // initialize hm with list of columns requested for output
> > +  		    	for (int i=0; i<columns.length; i++){ // if we are here we know columns is not null
> > +  		    		if (hm.containsKey(new String(getFamily((byte[])columns[i])))){
> > +  		    			hm.get(new String(getFamily((byte[])columns[i]))).add((byte[])columns[i]);
> > +  		    		}else{
> > +  		    			ArrayList<byte[]> al = new ArrayList<byte[]>();
> > +  		    			al.add((byte[])columns[i]);
> > +  		    			hm.put(new String(getFamily((byte[])columns[i])), al);
> > +  		    		}	    			
> > +  		    	}
> > +  		    	
> > +  	    	if (hm.size()==1){//only one column family
> > +  	    		resultColumnsOnlyFilter = new FilterList(FilterList.Operator.MUST_PASS_ALL);
>
> Thank you for the explanation!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/255/files#r49526031>.
>



-- 
Regards, --Qifan